medusa/core/ensemble.py

```diff
@@ -72,52 +72,66 @@ def __init__(self,list_of_models=[], identifier=None, name=None):
+        all_reactions = set(rxn.id for rxn in base_model.reactions)
         for model in list_of_models:
-            new_reactions = set([rxn.id for rxn in model.reactions]) - \
-                                all_reactions
-            reactions_to_add = [model.reactions.get_by_id(rxn) for rxn in new_reactions]
-            base_model.add_reactions(reactions_to_add)
-            all_reactions = all_reactions | set([rxn.id for rxn in model.reactions])
-
+            model_rxn_ids = set(rxn.id for rxn in model.reactions)
+            new_reactions = model_rxn_ids - all_reactions
+            if new_reactions:
+                reactions_to_add = [model.reactions.get_by_id(rxn_id) for rxn_id in new_reactions]
+                base_model.add_reactions(reactions_to_add)
+                all_reactions.update(new_reactions)
         all_reactions = list(all_reactions)
 
-        # Determine reactions that vary in any model and construct a feature for
-        # each unique parameter value for that reaction in the ensemble
-        variable_reactions = []
+        # Pre-cache model reaction attributes in dicts to avoid repeated getattr calls
+        model_reaction_attrs = {}
+        for model in list_of_models:
+            mid = model.id
+            model_reaction_attrs[mid] = {}
+            for rxn in model.reactions:
+                # Direct attribute access instead of getattr
+                model_reaction_attrs[mid][rxn.id] = {
+                    'lower_bound': rxn.lower_bound,
+                    'upper_bound': rxn.upper_bound
+                }
+
+        # Iterate over each reaction to detect variable attributes
         for reaction in all_reactions:
+            # Collect reaction attributes per model
             rxn_vals = {}
             for model in list_of_models:
-                rxn_vals[model.id] = {}
-                if reaction in [x.id for x in model.reactions]:
-                    rxn = model.reactions.get_by_id(reaction)
-                    for reaction_attribute in REACTION_ATTRIBUTES:
-                        rxn_vals[model.id][reaction_attribute] = \
-                            getattr(rxn,reaction_attribute)
-                else: # for reactions not present in this model, select the default
-                    for reaction_attribute in REACTION_ATTRIBUTES:
-                        rxn_vals[model.id][reaction_attribute] = \
-                            MISSING_ATTRIBUTE_DEFAULT[reaction_attribute]
-
-            rxn_vals = pd.DataFrame(rxn_vals).T
+                mid = model.id
+                rxn_attrs = model_reaction_attrs[mid]
+                if reaction in rxn_attrs:
+                    rxn_vals[mid] = rxn_attrs[reaction]
+                else:
+                    # Use default bounds if reaction missing from model
+                    rxn_vals[mid] = {'lower_bound': 0, 'upper_bound': 0} # TODO remark: could use MISSING_ATTRIBUTE_DEFAULT instead
```

Comment:
@jgilis I think it's fine to keep your implementation as-is; the global REACTION_ATTRIBUTES and MISSING_ATTRIBUTE_DEFAULT were intended for generalizability anticipating implementation of alternative GPRs for the same reaction across ensemble members. The rest of Medusa doesn't support that yet, so I'm OK with these changes focusing the current implementation more towards reaction presence+directionality. Adding support for alternative GPRs will require some other refactoring and addition of utility functions for simulation methods that use GPRs anyway.